From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sun, 3 Aug 2025 20:38:56 -0500
Subject: [PATCH] Fish Schedule Async Tasks For End Of Tick

This implementation allows async threads to still do async operations as long as the world is not ticking.
If the world is ticking, the "request" will be scheduled to be run sync at the end of the tick.
  * Not doing this, means that if the world is always ticking (mspt > 50) there's a chance that the async thread will sleep indefinitely.
If the world is not ticking, the "request" will be performed async by the calling thread.
If the world is about to tick, but the async "request" is not done yet, the world will wait until its done and every new "request" will be scheduled.
This scheduling logic is very performant as long as none of the plugins is spamming world/block requests.

diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 6a871948798cd06b7b7bce08e761dfd7ee81d216..37675914a28b907c52344ec67f65b38133af82fe 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -229,6 +229,8 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     public boolean hasPhysicsEvent = true; // Paper - BlockPhysicsEvent
     public boolean hasEntityMoveEvent; // Paper - Add EntityMoveEvent
     private final alternate.current.wire.WireHandler wireHandler = new alternate.current.wire.WireHandler(this); // Paper - optimize redstone (Alternate Current)
+    public final java.util.Queue<Runnable> fish$endOfTickTasks = new java.util.concurrent.ConcurrentLinkedQueue<>(); // Fish - Schedule tasks for end of tick
+    public final java.util.concurrent.locks.ReentrantReadWriteLock fish$lock = new java.util.concurrent.locks.ReentrantReadWriteLock(true);
 
     @Override
     public @Nullable LevelChunk getChunkIfLoaded(int x, int z) {
